Map view types when converting arrow_expressions to vortex expressions

paultiq opened the pull request:
Fixes https://github.com/vortex-data/vortex/issues/5759

This PR maps string_views to strings and binary_views to binary so that `to_substrait` will no longer raise ArrowNotImplementedError when constructing the substrait. 

Vortex supports expressions over views and Arrow compute doesn't, but [to_substrait](https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Expression.html#pyarrow.dataset.Expression.to_substrait) raises ArrowNotImplementedError based on Arrow compute kernels... regardless of the backend.

paultiq committed:
fix: map view types to avoid to_substrait ArrowNotImplementedError

paultiq committed:
DCO Remediation Commit for Paul Timmins <paul@iqmo.com>

I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this commit: 8ae0c0473e79b009d4036969393588e664537bcf

Signed-off-by: Paul Timmins <paul@iqmo.com>

danking reviewed (approved):
This is great and we should merge! Can you follow up with a test that writes and reads through DuckDB a Vortex file containing strings? I think paultiq included one at one point in one of his GitHub issues.

paultiq commented:
@danking Do you want that test case in this PR or a separate? Happy to do either.

danking commented:
Separate is fine! Sorry I referred to you in the third person! GitHub mobile UI had me turned around haha

paultiq committed:
lint: fix lint error on test_expression (ruff check ---fix'd)

paultiq committed:
DCO Remediation Commit for Paul Timmins <paul@iqmo.com>

  I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this commit: de97c5b526d9d13de4e0efa60330cbb746312d26

  Signed-off-by: Paul Timmins <paul@iqmo.com>

paultiq committed:
DCO Remediation Commit for Paul Timmins <paul@iqmo.com>

I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this commit: de97c5b526d9d13de4e0efa60330cbb746312d26
I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this commit: 521482460cabb9c123d5480fae62e1f9c938ce7a

Signed-off-by: Paul Timmins <paul@iqmo.com>

danking committed:
Update test_expression.py

SPDX

joseph-isaacs commented:
@paultiq think we can get this over the line

paultiq commented:
> @paultiq think we can get this over the line

AFAIK, this is ready to merge, was there something else for this PR?

joseph-isaacs commented:
The CI / Python lint is failing, we need to fix that

paultiq committed:
Merge branch 'develop' into view_mapping_p

paultiq committed:
typechecking: add pyright: ignores for pyrights pyarrow warnings.
Signed-off-by: Paul Timmins <paul@iqmo.com>

joseph-isaacs commented:
Why did you ignore all the warnings?

paultiq commented:
> Why did you ignore all the warnings?

The root reason is lack of type stubs for pyarrow, which is being addressed in https://github.com/apache/arrow/issues/32609

The ignores are consistent with the existing code: 
https://github.com/vortex-data/vortex/blob/41bb2bf14f1582d6a34a2493c03ee4de62f12a4a/vortex-python/python/vortex/arrow/expression.py#L32-L35

And existing test cases: 
https://github.com/vortex-data/vortex/blob/41bb2bf14f1582d6a34a2493c03ee4de62f12a4a/vortex-python/benchmark/test_aggregation.py#L23-L24

(I'll fix that DCO in a sec...)

paultiq commented:
The one question would be whether the direct test of `_schema_for_substrait` is OK (hence the reportPrivateUsage). 
reportPrivateUsage is used in a few places in the project, although not in any test cases, so that'd be a judgement call. 

test_expression.py tests both paths: the "private" _schema_for_substrait, as well as the "public" arrow_to_vortex path.

So, the choice would be either to allow the `ignore: reportPrivateUsage` or drop the direct test of _schema_for_substrait.

paultiq committed:
DCO Remediation Commit for Paul Timmins <paul@iqmo.com>

I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this
commit: d80d54de2f29c771dc0ee0af958bf7be06f036f8

Signed-off-by: Paul Timmins <paul@iqmo.com>

paultiq committed:
DCO Remediation Commit for Paul Timmins <paul@iqmo.com>

I, Paul Timmins <paul@iqmo.com>, hereby add my Signed-off-by to this
commit: d53be9c7a03891c0715a93edca7a7c046fdde60a

Signed-off-by: Paul Timmins <paul@iqmo.com>

paultiq commented:
(sorry for the multiple DCO commits, just not in my normal habits)

The remaining DCO warning is for 3107e20814df357f36e08722eb88372eb6934550, @danking's commit.

danking commented:
Thank you @paultiq! Sorry for the friction in our process currently. I'll keep an eye on the pyarrow types issue, thanks for flagging.
